src/python/TaskWorker/Actions/RucioDataDiscovery.py

```diff
@@ -0,0 +1,467 @@
+                self.requestTapeRecall(blockList=blocksWithLocation, system='Rucio', msgHead=msg)
+            else:
+                msg = "Some blocks are on TAPE only and will not be processed."
+                msg += "\nThere is no automatic recall from tape for data tier %s" % dataTier
+                msg += '\nIf you need the full dataset, contact Data Transfer team via %s' % FEEDBACKMAIL
+                self.logger.warning(msg)
+                self.uploadWarning(msg, self.userproxy, self.taskName)
+
+        try:
+            # filedetails = self.dbs.listDatasetFileDetails(inputDataset, getParents=True, getLumis=needLumiInfo, validFileOnly=0)
+            # need a Rucio way to do the above, filedetails is {lfn1:detail1, lfn2:detail2,...}
+            # where for each file we have something like (from comments in DBS3Reader.py)
+            #             { 'NumberOfEvents': 545,
+            #               'BlockName': '/HighPileUp/Run2011A-v1/RAW#dd6e0796-cbcc-11e0-80a9-003048caaace',
+            #               'Lumis': {173658: [8, 12, 9, 14, 19, 109, 105]},
+            #               'Parents': [],
+            #               'Checksum': '22218315',
+            #               'Adler32': 'a41a1446',
+            #               'Md5' : '23415' , # not listed in DBS3Reaser comments, but rquired in WMCore pff...
+            #               'FileSize': 286021145,
+            #               'ValidFile': 1
+            #             }
+            # since RucioDataDiscovery only supports FileBased splitting, we can put dummy values for lumis
+            # And yes, there's something perverse in starting from blocks to create flat list of files with
+            # block name in each so that eventually splitter can manipulate those files... still respecting
+            # block boundaries. huh ? I (SB) do not know how something like this could have been devised
+            # in first place. If anything else the work in formatOutput is way more then it ought to be
+            # Maybe by now they have changed how they do things in WMA, so we could align, or else, lead the way.
+            #
+            # Here's DBS3Reader.listDatsetFileDetails re-implemented via Rucio API
+            filedetails = {}
+            detail = {  # a template
+                'NumberOfEvents': 0, 'BlockName': 'to be filled', 'Lumis': {}, 'Parents': [],
+                'Checksum': 'aChecksum', 'Adler32': 'anAdler', 'Md5': 'anMd5',
+                'FileSize': 123456789, 'ValidFile': 1}
+            for dataset in blocksWithLocation:  #  DBS block = Rucio dataset
+                detail['BlockName'] = dataset
+                dids = self.rucioClient.list_files(rucioScope, dataset)
+                for did in dids:
+                    lfn = did['name']
+                    detail['FileSize'] = did['bytes']
+                    detail['Adler32'] = did['adler32']
+                    filedetails[lfn] = detail
```

Comment:
Do `detail` objects in `filedetails` for each key is the same?
Let say when you access `filedetails['file1.root']['FileSize']` and `filedetails['file2.root']['FileSize']` you will get the same value.

Reply:
I think not. for every LFN `detail` gets new values and is then added as  to `filedetails` as `filedetails['file2']` . This code should be pretty much the same as in DBSDataDiscovery . I will look into using some common function here as well

Reply:
But it is the same object for every LFN. Unless you put declare object `detail = {...}` inside first loop, no?

Reply:
Yes. `copy` is fine for this case, and we need to put in the second loop (not the first loop as I said earlier).
Keep in mind that `copy` will not copy the values, only copy the pointer.

Reply:
let's make sure I understand, something like this ?
```python
from copy import deepcopy
filedetails={}  # dict. to fill
detail=aTemplateDictionary
for dataset in list-of-datasets:
    detail[someKey]=someValue
    for did in list-of-dids:
       detail[anotherKey=anotherValue]
       filedetails[lfn] = deepcopy(detail)
```
all in all, when preparing a `filedetails` dict objec to be handed to downstream code which does I-do-not-know-what with it.. better not have pointers in there to things which I define in here !

Reply:
resolved via https://github.com/dmwm/CRABServer/pull/7556/commits/1a68b84ff390c9d393961083d998234f639d2762